scalding-parquet-scrooge/src/main/scala/com/twitter/scalding/parquet/scrooge/ParquetCollectionFormatForwardCompatibility.scala

```
@@ -0,0 +1,130 @@
+ * schema in terms of field names. However, the data types for collection can differ in
+ * graph structure between the two schemas. We then need to:
+ * 1) traverse the two schemas until we find the collection type indicated by `repeated` type.
+ * 2) delegate the collection types found to respective list/map formatter.
+ */
+private[scrooge] object ParquetCollectionFormatForwardCompatibility {
+
+  private val logger = LoggerFactory.getLogger(getClass)
+
+  /**
+   * Create a forward-compatible schema, using content from source type with format from target type.
+   *
+   * @param sourceType source type with legacy format
+   * @param targetType target type to which source is converted to
+   */
+  def formatForwardCompatibleMessage(sourceType: MessageType, targetType: MessageType): MessageType = {
+    val groupResult = formatForwardCompatibleType(sourceType, targetType).asGroupType()
+    logger.debug("Making source schema to be compatible with target" +
+      s"\nSource:\n${sourceType}\nTarget:\n${targetType}\nResult:\n${groupResult}")
+    new MessageType(groupResult.getName, groupResult.getFields)
+  }
+
+  /**
+   * Traverse source/target schemas and format nodes of list or map.
+   * The formatting is not to one-to-one node swapping from source to target,
+   * this is because the subset fields of source node and its optional/required must
+   * be maintained in the formatted result.
+   */
+  private def formatForwardCompatibleType(sourceType: Type, targetType: Type): Type = {
+    (findCollectionGroup(sourceType), findCollectionGroup(targetType)) match {
+      case (Some(sourceGroup: ListGroup), Some(targetGroup: ListGroup)) =>
+        formatForwardCompatibleCollectionGroup[ListGroup](sourceGroup, targetGroup)
+      case (Some(sourceGroup: MapGroup), Some(targetGroup: MapGroup)) =>
+        formatForwardCompatibleCollectionGroup[MapGroup](sourceGroup, targetGroup)
+      case _ if sourceType.isPrimitive || targetType.isPrimitive => // Base case
```

I would explicitly check and throw if one is a primitive but not the other. As written this will silently pass through certain schema mismatches.

---

Ack